connectors/sources/github.py

```diff
@@ -0,0 +1,252 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""GitHub source module responsible to fetch documents from GitHub Cloud and Server."""
+import json
+from functools import cached_property
+
+import aiohttp
+from gidgethub import BadRequest
+from gidgethub.aiohttp import GitHubAPI
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource, ConfigurableFieldValueError
+from connectors.utils import CancellableSleeps, RetryStrategy, iso_utc, retryable
+
+RETRIES = 3
+RETRY_INTERVAL = 2
+
+USER_API = "https://api.github.com/user"
+ENDPOINTS = {
+    "ALL_REPOS": "/user/repos?type=all",
+    "REPO": "repos/{repo_name}",
+    "USER": "/user",
+}
+
+
+class GitHubClient:
+    """GitHub Client"""
+
+    def __init__(self, configuration):
+        """Setup the GitHub client.
+
+        Args:
+            configuration (DataSourceConfiguration): Instance of DataSourceConfiguration class.
+        """
+
+        self.session = None
+        self._sleeps = CancellableSleeps()
+        self.configuration = configuration
+        self.repos = self.configuration["repositories"]
+        self.github_token = self.configuration["github_token"]
+        self.headers = {"Authorization": f"Bearer {self.github_token}"}
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def execute_getitem(self, request):
```

Comment:
```suggestion
    async def get_item(self, resource):
```

As we're passing a URI I think `request` is a misleading name.

We could also use an enumeration for the different resources + a dictionary:

```
class GitHubResource(Enum):
   USER = ...
   REPOSITORY = ...
   
URIS = {
  REPOSITORY: {
     ALL: "/user/repos?type=all",
     BY_NAME: "repos/{repo_name}"
  }
}
```

Feel free to use this approach or just rename the method.

Reply:
We think enum become complex so currently rename the method